In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
#import os
#current_dir = os.getcwd()
#os.getcwd() test

## Import dataset
<p>Premièrement, on va importer les fichiers CSV des entreprises, de la géographie, des salaires et de la population et afficher l’aperçu.<p>


In [17]:
# Import and review the data 

entreprise = pd.read_csv('../data/base_etablissement_par_tranche_effectif.csv') 
display(entreprise.head())

salary = pd.read_csv('../data/net_salary_per_town_categories.csv')
display(salary.head())

name_geographic = pd.read_csv('../data/name_geographic_information.csv') 
display(name_geographic.head())

population = pd.read_csv('../data/population.csv', dtype={'CODGEO': object} )
display(population.head())

CODGEO                   LIBGEO  REG DEP  E14TST  E14TS0ND  E14TS1  E14TS6  \
0  01001  L'Abergement-Clémenciat   82  01      25        22       1       2   
1  01002    L'Abergement-de-Varey   82  01      10         9       1       0   
2  01004        Ambérieu-en-Bugey   82  01     996       577     272      63   
3  01005      Ambérieux-en-Dombes   82  01      99        73      20       3   
4  01006                  Ambléon   82  01       4         4       0       0   

   E14TS10  E14TS20  E14TS50  E14TS100  E14TS200  E14TS500  
0        0        0        0         0         0         0  
1        0        0        0         0         0         0  
2       46       24        9         3         2         0  
3        1        2        0         0         0         0  
4        0        0        0         0         0         0

CODGEO             LIBGEO  SNHM14  SNHMC14  SNHMP14  SNHME14  SNHMO14  \
0  01004  Ambérieu-en-Bugey    13.7     24.2     15.5     10.3     11.2   
1  01007           Ambronay    13.5     22.1     14.7     10.7     11.4   
2  01014             Arbent    13.5     27.6     15.6     11.1     11.1   
3  01024           Attignat    12.9     21.8     14.1     11.0     11.3   
4  01025      Bâgé-la-Ville    13.0     22.8     14.1     10.5     11.1   

   SNHMF14  SNHMFC14  SNHMFP14  ...  SNHMHO14  SNHM1814  SNHM2614  SNHM5014  \
0     11.6      19.1      13.2  ...      11.6      10.5      13.7      16.1   
1     11.9      19.0      13.3  ...      11.7       9.8      13.8      14.6   
2     10.9      19.5      11.7  ...      11.8       9.3      13.3      16.0   
3     11.4      19.0      13.0  ...      11.6       9.6      12.9      14.2   
4     11.6      19.4      13.6  ...      11.4       9.4      12.8      15.2   

   SNHMF1814  SNHMF2614  SNHMF5014  SNHMH1814  SNHMH2614  SNHMH5014  
0        9.7       11.8       12.5       11.0       14.9       18.6  
1        9.2       12.2       12.5       10.2       14.9       16.4  
2        8.9       10.6       12.5        9.6       15.1       18.6  
3        9.3       11.4       12.2        9.7       13.8       15.9  
4        9.0       11.8       12.3        9.7       13.4       16.9  

[5 rows x 26 columns]

EU_circo  code_région   nom_région chef.lieu_région numéro_département  \
0  Sud-Est           82  Rhône-Alpes             Lyon                 01   
1  Sud-Est           82  Rhône-Alpes             Lyon                 01   
2  Sud-Est           82  Rhône-Alpes             Lyon                 01   
3  Sud-Est           82  Rhône-Alpes             Lyon                 01   
4  Sud-Est           82  Rhône-Alpes             Lyon                 01   

  nom_département       préfecture  numéro_circonscription  \
0             Ain  Bourg-en-Bresse                       1   
1             Ain  Bourg-en-Bresse                       1   
2             Ain  Bourg-en-Bresse                       1   
3             Ain  Bourg-en-Bresse                       1   
4             Ain  Bourg-en-Bresse                       1   

            nom_commune codes_postaux  code_insee   latitude longitude  \
0              Attignat         01340        1024  46.283333  5.166667   
1              Beaupont         01270        1029  46.400000  5.266667   
2                  Bény         01370        1038  46.333333  5.283333   
3             Béreyziat         01340        1040  46.366667      5.05   
4  Bohas-Meyriat-Rignat         01250        1245  46.133333       5.4   

   éloignement  
0         1.21  
1         1.91  
2         1.51  
3         1.71  
4         1.01

NIVGEO CODGEO                   LIBGEO  MOCO  AGEQ80_17  SEXE  NB
0    COM  01001  L'Abergement-Clémenciat    11          0     1  15
1    COM  01001  L'Abergement-Clémenciat    11          0     2  15
2    COM  01001  L'Abergement-Clémenciat    11          5     1  20
3    COM  01001  L'Abergement-Clémenciat    11          5     2  20
4    COM  01001  L'Abergement-Clémenciat    11         10     1  20

In [24]:
# Review data types and columns info
print(entreprise.info())
print(salary.info())
print(name_geographic.info())
print(population.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36681 entries, 0 to 36680
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   CODGEO    36681 non-null  object
 1   LIBGEO    36681 non-null  object
 2   REG       36681 non-null  int64 
 3   DEP       36681 non-null  object
 4   E14TST    36681 non-null  int64 
 5   E14TS0ND  36681 non-null  int64 
 6   E14TS1    36681 non-null  int64 
 7   E14TS6    36681 non-null  int64 
 8   E14TS10   36681 non-null  int64 
 9   E14TS20   36681 non-null  int64 
 10  E14TS50   36681 non-null  int64 
 11  E14TS100  36681 non-null  int64 
 12  E14TS200  36681 non-null  int64 
 13  E14TS500  36681 non-null  int64 
dtypes: int64(11), object(3)
memory usage: 3.9+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5136 entries, 0 to 5135
Data columns (total 26 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   CODGEO     5136 non-null   object 


In [64]:
# Shape for each DataFrame 

print("Shape of DF:")
print("entreprise:",entreprise.shape)
print("salary:",salary.shape)
print("name_geographic:",name_geographic.shape)
print("population:",population.shape)


Shape of DF:
entreprise: (36681, 14)
salary: (5136, 26)
name_geographic: (36840, 14)
population: (8536584, 7)


In [62]:
# Number of duplicates for each DataFrame

print('Number of duplicates:')
print('entreprise:', entreprise.duplicated().sum())
print('salary:', salary.duplicated().sum())
print('name_geographic:', name_geographic.duplicated().sum())
print('population:', population.duplicated().sum())


Number of duplicates:
entreprise: 0
salary: 0
name_geographic: 0
population: 0


In [35]:
# Function that provides a short summary of DataFrame with information about the number of unique values, 
# percentage of missing values, number of missing values, and the data type of each column.

def summary_short(df):
    """
    Function that provides a short summary of DataFrame with information about the number of unique values, 
    percentage of missing values, number of missing values, and the data type of each column.
    
    Is used with not cleaned DataFrames.
    
    Parameters:
    df (DataFrame): The DataFrame for which the summary is required.
    
    Returns:
    DataFrame with 
    """
    
    tab = pd.DataFrame(index = df.columns, columns = ['nb_unique_values','%_missing_values','nb_missing_values','type'])
    tab.loc[:,'nb_unique_values'] = df.nunique().values
    tab.loc[:,'%_missing_values'] = df.isna().sum().values / len(df)
    tab.loc[:,'nb_missing_values'] = df.isna().sum().values
    tab.loc[:, 'type'] = df.dtypes.values
    
    return display(tab)

In [36]:
summary_short(name_geographic)

nb_unique_values %_missing_values nb_missing_values  \
EU_circo                              8              0.0                 0   
code_région                          28              0.0                 0   
nom_région                           28              0.0                 0   
chef.lieu_région                     28              0.0                 0   
numéro_département                  102              0.0                 0   
nom_département                     102              0.0                 0   
préfecture                          102              0.0                 0   
numéro_circonscription               24              0.0                 0   
nom_commune                       34142              0.0                 0   
codes_postaux                      6106              0.0                 0   
code_insee                        36693              0.0                 0   
latitude                            799         0.079506              2929   
longitude                          1150         0.077117              2841   
éloignement                         293         0.080402              2962   

                           type  
EU_circo                 object  
code_région               int64  
nom_région               object  
chef.lieu_région         object  
numéro_département       object  
nom_département          object  
préfecture               object  
numéro_circonscription    int64  
nom_commune              object  
codes_postaux            object  
code_insee                int64  
latitude                float64  
longitude                object  
éloignement             float64

In [40]:
#population['CODGEO'].astype(float)

# Data type verification  
def detection_error(element):
    try:
        float(element)
    except ValueError:
        return element
    
print(name_geographic.longitude.apply(detection_error).loc[name_geographic.longitude.apply(detection_error).notna()].values)


['5,83' '6,01' '6,24' '5,82' '6,31' '4,77' '0,94' '2,07' '0,97' '1,54'
 '1,67' '1,5' '1,78' '1,69' '1,8' '1,63' '1,8' '1,88' '1,67' '0,89' '1,21'
 '1,23' '0,98' '1,15' '1,31' '1,09' '1,18' '1,5' '1,38' '1,11' '1,71'
 '1,93' '1,17' '1,12' '1,68' '1,64' '1,28' '1,02' '1,17' '1,5' '2,25'
 '2,03' '2,15' '2,43' '3,09' '2,89' '4,76' '-' '0,08' '-' '-' '-' '-' '-'
 '-' '-' '-' '-' '-' '-' '-' '-' '2,51' '4,33' '-2,24' '-3,43' '1,91'
 '1,3' '-5,1' '1,29' '1,2' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-'
 '-' '5,28' '5,28' '5,96' '-' '1,74' '4,07' '3,89' '3,89' '-' '-' '-'
 '3,48' '3,47' '3,93' '2,93' '3,93' '3,84' '-' '-' '-' '-' '-' '-' '-'
 '1,76' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-'
 '-' '3,01' '2,03' '4,9' '4,32' '-' '-' '-' '5,84' '2,55' '-' '-' '-'
 '-0,75' '1,95' '1,97' '6,74' '5,04' '-1,94' '0,25' '-' '0,41' '-' '-'
 '3,28']


## Conclusion pour name_geographic DF : 
- Les noms des colonnes en français, tandis que ceux des autres DataFrames sont en anglais. Il est donc nécessaire de les renommer pour une cohérence.
- La colonne "code_insee" semble correspondre à "CODGEO", mais il y a un '0' au debut, par example, '1024' au lieu de '01024'.
- Les colonnes "latitude", "longitude", "éloignement" présentent respectivement 2929, 2841 et 2962 valeurs manquantes (NaN).
- La colonne "longitude" devrait normalement être de type float64. Cependant, nous n'avons pas pu modifier le type en raison de problèmes avec les données. Par exemple :
  
    - L'utilisation de ',' au lieu de '.' pour la notation décimale : **'5,83'**
    - La présence de '-' : ceux-ci doivent être remplacés par des valeurs NaN.


In [55]:
def summary(df):
    """
    Provides a summary of the DataFrame with information about the number of unique values, 
    percentage of missing values, data type of each column, mean or mode depending on the data type,
    and potential alerts such as too many missing values or imbalanced categories.
    
    Parameters:
    df (DataFrame): The DataFrame for which the summary is required.
    
    Returns:
    None
    """
    
    table = pd.DataFrame(
        index=df.columns,
        columns=['type_info', '%_missing_values', 'nb_unique_values', 'nb_zero_values','%_zero_values', 'list_unique_values', "mean_or_mode", "flag"])
    
    # Fill in column 'type_info' with data types of each column
    table.loc[:, 'type_info'] = df.dtypes.values
    
    # Calculate the percentage of missing values for each column
    table.loc[:, '%_missing_values'] = np.round((df.isna().sum().values / len(df)) * 100)
    
    # Calculate the number of unique values for each column
    table.loc[:, 'nb_unique_values'] = df.nunique().values

    # Calculate the number of 0 values for each column
    table.loc[:, 'nb_zero_values'] = df.isin([0]).sum().values

    # Calculate the percentage of 0 values for each column
    table.loc[:, '%_zero_values'] = np.round((df.isin([0]).sum().values/len(df)) * 100)
    
    # Function to get the list of unique values for a column, showing "Too many categories..." if there are too many
    def get_list_unique_values(colonne):
        if colonne.nunique() < 11:
            return colonne.unique()
        else:
            return "Too many categories..." if colonne.dtypes == "O" else "Too many values..."
    
    # Function to get the mean or mode depending on the data type
    def get_mean_mode(colonne):
        if (colonne.dtypes == "O") or (len(colonne.unique()) < 20):
            return colonne.astype('category').mode()[0]
        else: 
            return colonne.mean()
    
    # Function to check for potential alerts such as too many missing values or imbalanced categories
    def alerts(colonne, thresh_na=0.25, thresh_balance=0.8):
        if (colonne.isna().sum()/len(df)) > thresh_na:
            return "Too many missing values!"
        elif colonne.value_counts(normalize=True).values[0] > thresh_balance:
            return "It's imbalanced!"
        else:
            return "Nothing to report"
        
    # Fill in the columns with the respective information
    table.loc[:, 'list_unique_values'] = df.apply(get_list_unique_values)
    table.loc[:, 'mean_or_mode'] = df.apply(get_mean_mode)
    table.loc[:, 'flag'] = df.apply(alerts)
    
    # Display the summary table
    return display(table)

In [59]:
summary(entreprise)

type_info %_missing_values nb_unique_values nb_zero_values  \
CODGEO      object              0.0            36681              0   
LIBGEO      object              0.0            34142              0   
REG          int64              0.0               27              0   
DEP         object              0.0              101              0   
E14TST       int64              0.0             1423            399   
E14TS0ND     int64              0.0             1125            579   
E14TS1       int64              0.0              650           6118   
E14TS6       int64              0.0              272          20324   
E14TS10      int64              0.0              220          22641   
E14TS20      int64              0.0              160          25884   
E14TS50      int64              0.0               87          30678   
E14TS100     int64              0.0               57          33108   
E14TS200     int64              0.0               42          34460   
E14TS500     int64              0.0               23          35933   

         %_zero_values      list_unique_values    mean_or_mode  \
CODGEO             0.0  Too many categories...           01001   
LIBGEO             0.0  Too many categories...  Sainte-Colombe   
REG                0.0      Too many values...       49.415365   
DEP                0.0  Too many categories...              62   
E14TST             1.0      Too many values...      123.456067   
E14TS0ND           2.0      Too many values...       83.555301   
E14TS1            17.0      Too many values...       27.291486   
E14TS6            55.0      Too many values...         5.22055   
E14TS10           62.0      Too many values...        3.800333   
E14TS20           71.0      Too many values...        2.296448   
E14TS50           84.0      Too many values...        0.738339   
E14TS100          90.0      Too many values...        0.332434   
E14TS200          94.0      Too many values...         0.17276   
E14TS500          98.0      Too many values...        0.048417   

                       flag  
CODGEO    Nothing to report  
LIBGEO    Nothing to report  
REG       Nothing to report  
DEP       Nothing to report  
E14TST    Nothing to report  
E14TS0ND  Nothing to report  
E14TS1    Nothing to report  
E14TS6    Nothing to report  
E14TS10   Nothing to report  
E14TS20   Nothing to report  
E14TS50    It's imbalanced!  
E14TS100   It's imbalanced!  
E14TS200   It's imbalanced!  
E14TS500   It's imbalanced!

## Conclusion pour name_geographic DF : 
- Il y a beaucoup de zeros dans la DF pour les colonnes suivantes :
    - 'E14TS6', 'E14TS10', 'E14TS20', 'E14TS50', 'E14TS100', 'E14TS200',
       'E14TS500'
- Pour améliorer cela, nous pouvons créer de nouvelles colonnes qui catégorisent les données en fonction de tailles d'entreprises plus vastes :
    - Micro entreprise: 0 <= taille < 10 personnes
    - Petit entreprise: 10 <= taille < 50 personnes
    - Moyenne entreprise :  50 <= taille < 200 personnes
    - Grande Entreprise : taille >= 200 personnes

In [60]:
summary(salary)

type_info %_missing_values nb_unique_values nb_zero_values  \
CODGEO       object              0.0             5136              0   
LIBGEO       object              0.0             5085              0   
SNHM14      float64              0.0              167              0   
SNHMC14     float64              0.0              202              0   
SNHMP14     float64              0.0               99              0   
SNHME14     float64              0.0               69              0   
SNHMO14     float64              0.0               97              0   
SNHMF14     float64              0.0              129              0   
SNHMFC14    float64              0.0              175              0   
SNHMFP14    float64              0.0               74              0   
SNHMFE14    float64              0.0               59              0   
SNHMFO14    float64              0.0              103              0   
SNHMH14     float64              0.0              200              0   
SNHMHC14    float64              0.0              228              0   
SNHMHP14    float64              0.0              122              0   
SNHMHE14    float64              0.0               98              0   
SNHMHO14    float64              0.0               96              0   
SNHM1814    float64              0.0               62              0   
SNHM2614    float64              0.0              160              0   
SNHM5014    float64              0.0              224              0   
SNHMF1814   float64              0.0               43              0   
SNHMF2614   float64              0.0              128              0   
SNHMF5014   float64              0.0              154              0   
SNHMH1814   float64              0.0               72              0   
SNHMH2614   float64              0.0              192              0   
SNHMH5014   float64              0.0              266              0   

          %_zero_values      list_unique_values  mean_or_mode  \
CODGEO              0.0  Too many categories...         01004   
LIBGEO              0.0  Too many categories...  Sainte-Marie   
SNHM14              0.0      Too many values...     13.706386   
SNHMC14             0.0      Too many values...     23.703836   
SNHMP14             0.0      Too many values...     14.575876   
SNHME14             0.0      Too many values...     10.564505   
SNHMO14             0.0      Too many values...      11.23565   
SNHMF14             0.0      Too many values...     12.038026   
SNHMFC14            0.0      Too many values...     20.220502   
SNHMFP14            0.0      Too many values...      13.27426   
SNHMFE14            0.0      Too many values...     10.308879   
SNHMFO14            0.0      Too many values...      9.827161   
SNHMH14             0.0      Too many values...     14.848189   
SNHMHC14            0.0      Too many values...     25.196963   
SNHMHP14            0.0      Too many values...     15.493789   
SNHMHE14            0.0      Too many values...     11.265713   
SNHMHO14            0.0      Too many values...     11.498189   
SNHM1814            0.0      Too many values...      9.549766   
SNHM2614            0.0      Too many values...     13.495444   
SNHM5014            0.0      Too many values...      15.87722   
SNHMF1814           0.0      Too many values...      9.161565   
SNHMF2614           0.0      Too many values...      12.05516   
SNHMF5014           0.0      Too many values...     13.174143   
SNHMH1814           0.0      Too many values...      9.820872   
SNHMH2614           0.0      Too many values...     14.485981   
SNHMH5014           0.0      Too many values...     17.679809   

                        flag  
CODGEO     Nothing to report  
LIBGEO     Nothing to report  
SNHM14     Nothing to report  
SNHMC14    Nothing to report  
SNHMP14    Nothing to report  
SNHME14    Nothing to report  
SNHMO14    Nothing to report  
SNHMF14    Nothing to report  
SNHMFC14   Nothing to report  

In [65]:
# Descriptive statistics for salary dataset

salary.describe()

SNHM14      SNHMC14      SNHMP14      SNHME14      SNHMO14  \
count  5136.000000  5136.000000  5136.000000  5136.000000  5136.000000   
mean     13.706386    23.703836    14.575876    10.564505    11.235650   
std       2.559329     2.836183     1.490110     0.811775     1.221755   
min      10.200000    16.000000    11.600000     8.700000     8.300000   
25%      12.100000    21.900000    13.800000    10.000000    10.600000   
50%      13.000000    23.200000    14.400000    10.400000    11.000000   
75%      14.400000    24.900000    15.100000    10.900000    11.600000   
max      43.300000    51.500000    54.600000    17.500000    46.300000   

           SNHMF14     SNHMFC14     SNHMFP14     SNHMFE14     SNHMFO14  ...  \
count  5136.000000  5136.000000  5136.000000  5136.000000  5136.000000  ...   
mean     12.038026    20.220502    13.274260    10.308879     9.827161  ...   
std       1.787995     2.327550     0.990166     0.747563     1.104684  ...   
min       9.300000    12.000000    10.600000     8.700000     6.100000  ...   
25%      10.900000    18.800000    12.600000     9.800000     9.200000  ...   
50%      11.500000    20.000000    13.100000    10.100000     9.700000  ...   
75%      12.700000    21.400000    13.800000    10.600000    10.200000  ...   
max      26.700000    35.500000    19.000000    16.100000    28.100000  ...   

          SNHMHO14     SNHM1814     SNHM2614    SNHM5014    SNHMF1814  \
count  5136.000000  5136.000000  5136.000000  5136.00000  5136.000000   
mean     11.498189     9.549766    13.495444    15.87722     9.161565   
std       1.289598     0.997444     2.363144     3.58586     0.453023   
min       8.900000     7.900000     9.700000    10.50000     7.500000   
25%      10.800000     9.200000    12.000000    13.70000     8.900000   
50%      11.300000     9.500000    12.900000    15.00000     9.100000   
75%      11.900000     9.700000    14.300000    16.90000     9.400000   
max      53.200000    60.600000    38.100000    56.90000    12.000000   

         SNHMF2614    SNHMF5014    SNHMH1814    SNHMH2614    SNHMH5014  
count  5136.000000  5136.000000  5136.000000  5136.000000  5136.000000  
mean     12.055160    13.174143     9.820872    14.485981    17.679809  
std       1.825306     2.249391     1.539949     2.852490     4.530257  
min       9.100000     9.500000     7.800000     9.600000    10.800000  
25%      10.900000    11.700000     9.400000    12.700000    14.900000  
50%      11.600000    12.600000     9.700000    13.800000    16.600000  
75%      12.700000    14.000000    10.000000    15.500000    19.000000  
max      26.600000    31.000000    93.300000    45.400000    68.600000  

[8 rows x 24 columns]

## Conclusion pour salary DF : 

- Il y a de nombreuses catégories de salariés, mais on peut voir qu'elles sont divisées selon les catégories suivantes : catégorie d'emploi, sexe et âge. Ça peut faciliter analyse. 

In [61]:
summary(population)

type_info %_missing_values nb_unique_values nb_zero_values  \
NIVGEO       object              0.0                1              0   
CODGEO       object              0.0            35868              0   
LIBGEO       object              0.0            33452              0   
MOCO          int64              0.0                7              0   
AGEQ80_17     int64              0.0               17         502152   
SEXE          int64              0.0                2              0   
NB            int64              0.0             2953        6325631   

          %_zero_values            list_unique_values    mean_or_mode  \
NIVGEO              0.0                         [COM]             COM   
CODGEO              0.0        Too many categories...           01001   
LIBGEO              0.0        Too many categories...  Sainte-Colombe   
MOCO                0.0  [11, 12, 21, 22, 23, 31, 32]              11   
AGEQ80_17           6.0            Too many values...               0   
SEXE                0.0                        [1, 2]               1   
NB                 74.0            Too many values...        7.446743   

                        flag  
NIVGEO      It's imbalanced!  
CODGEO     Nothing to report  
LIBGEO     Nothing to report  
MOCO       Nothing to report  
AGEQ80_17  Nothing to report  
SEXE       Nothing to report  
NB         Nothing to report

## Conclusion pour population DF :
- NIVGEO : une seule valeur unique (COM) est présente
- NB : Il y a un grand nombre de zéros (~74%). Il faut faire attention lors de l'analyse. Peut-être supprimer au future.
- AGEQ80_17 : Il y a un grand nombre de zéros, mais cela semble correct, 0 représente les personnes âgées de 0 à 4 ans. 
